Source/AMRInterpolator/SurfaceExtraction.impl.hpp

```diff
@@ -0,0 +1,411 @@
+{
+    if (a_interpolator == nullptr)
+    {
+        MayDay::Error("SurfaceExtraction: invalid AMRInterpolator pointer");
+    }
+    // set up the interpolation query
+    InterpolationQuery query(m_num_points * m_params.num_surfaces);
+    FOR1(idir) { query.setCoords(idir, m_interp_coords[idir].data()); }
+    for (int ivar = 0; ivar < m_vars.size(); ++ivar)
+    {
+        query.addComp(m_vars[ivar].first, m_interp_data[ivar].data(),
+                      m_vars[ivar].second);
+    }
+
+    // submit the query
+    a_interpolator->interp(query);
+    m_done_extraction = true;
+}
+
+//! Add an integrand (which must of type integrand_t) for integrate() to
+//! integrate over. Note the area_element is already included from the
+//! SurfaceGeometry template class
+template <class SurfaceGeometry>
+void SurfaceExtraction<SurfaceGeometry>::add_integrand(
+    const integrand_t &a_integrand, std::vector<double> &out_integrals,
+    const IntegrationMethod &a_method_u, const IntegrationMethod &a_method_v)
+{
+    // store the integrand
+    m_integrands.push_back(a_integrand);
+
+    // resize the out_integrals and store a reference to it
+    out_integrals.resize(m_params.num_surfaces);
+    std::fill(out_integrals.begin(), out_integrals.end(), 0.0);
+    m_integrals.push_back(std::ref(out_integrals));
+
+    // check if integration methods are valid given periodicity and number of
+    // points
+    bool valid_u =
+        a_method_u.is_valid(m_params.num_points_u, m_geom.is_u_periodic());
+    bool valid_v =
+        a_method_v.is_valid(m_params.num_points_v, m_geom.is_v_periodic());
+
+    // default to using the trapezium rule if provided methods are not valid
+    IntegrationMethod method_u = IntegrationMethod::trapezium;
+    IntegrationMethod method_v = IntegrationMethod::trapezium;
+    if (!valid_u)
+    {
+        MayDay::Warning(
+            "SurfaceExtraction<SurfaceGeometry>::integrate: Provided "
+            "IntegrationMethod for u is not valid with\nthis num_points_u; "
+            "reverting to trapezium rule.");
+    }
+    else
+    {
+        method_u = a_method_u;
+    }
+    if (!valid_v)
+    {
+        MayDay::Warning(
+            "SurfaceExtraction<SurfaceGeometry>::integrate: Provided "
+            "IntegrationMethod for v is not valid with\nthis num_points_v; "
+            "reverting to trapezium rule.");
+    }
+    else
+    {
+        method_v = a_method_v;
+    }
+    m_integration_methods.push_back({method_u, method_v});
+}
+
+//! Add an integrand which is just a single var. The a_var argument should
+//! correspond to the order in which the desired var was added to this object
+//! with add_var
+template <class SurfaceGeometry>
+void SurfaceExtraction<SurfaceGeometry>::add_var_integrand(
+    int a_var, std::vector<double> &out_integrals,
+    const IntegrationMethod &a_method_u, const IntegrationMethod &a_method_v)
+{
+    CH_assert(a_var > 0 && a_var < m_vars.size());
```

Comment:
I think this assertion is wrong. It should be `a_var >= 0`. I was getting wrong assertions in the example I built and couldn't understand why :P

Reply:
Yes I agree (or perhaps I was implicitly banning the extraction of `chi` :laughing: ).

Reply:
Yes I agree.